x-pack/plugins/files/public/components/upload_file/upload_state.ts

```diff
@@ -0,0 +1,189 @@
+  }
+
+  private validateFiles(files: File[]): undefined | string {
+    if (
+      this.fileKind.maxSizeBytes != null &&
+      files.some((file) => file.size > this.fileKind.maxSizeBytes!)
+    ) {
+      return i18nTexts.fileTooLarge(String(this.fileKind.maxSizeBytes));
+    }
+    return;
+  }
+
+  public setFiles = (files: File[]): void => {
+    if (this.isUploading()) {
+      throw new Error('Cannot update files while uploading');
+    }
+
+    const validationError = this.validateFiles(files);
+
+    this.files$$.next(
+      files.map((file) =>
+        createStateSubject<FileState>({
+          file,
+          status: 'idle',
+          error: validationError ? new Error(validationError) : undefined,
+        })
+      )
+    );
+  };
+
+  public abort = (): void => {
+    if (!this.isUploading()) {
+      throw new Error('No upload in progress');
+    }
+    this.abort$.next();
+  };
+
+  /**
+   * Do not throw from this method, it is intended to work with {@link forkJoin} from rxjs which
+   * unsubscribes from all observables if one of them throws.
+   */
+  private uploadFile = (
+    file$: SimpleStateSubject<FileState>,
+    abort$: Observable<void>,
+    meta?: unknown
+  ): Observable<void | Error> => {
+    const abortController = new AbortController();
+    const abortSignal = abortController.signal;
+    const { file, status } = file$.getValue();
+    if (!['idle', 'upload_failed'].includes(status)) {
+      return of(undefined);
+    }
+
+    let uploadTarget: undefined | FileJSON;
+    let erroredOrAborted = false;
+
+    file$.setState({ status: 'uploading', error: undefined });
+
+    const { name, mime } = parseFileName(file.name);
+
+    return from(
+      this.client.create({
+        kind: this.fileKind.id,
+        name,
+        mimeType: mime,
+        meta: meta as Record<string, unknown>,
+      })
+    ).pipe(
+      mergeMap((result) => {
+        uploadTarget = result.file;
+        return race(
+          abort$.pipe(
+            map(() => {
+              abortController.abort();
+              throw new Error('Abort!');
+            })
+          ),
+          this.client.upload({
+            body: file,
+            id: uploadTarget.id,
+            kind: this.fileKind.id,
+            abortSignal,
+          })
+        );
+      }),
+      map(() => {
+        file$.setState({ status: 'uploaded', id: uploadTarget?.id });
+      }),
+      catchError((e) => {
+        erroredOrAborted = true;
+        file$.setState({ status: 'upload_failed', error: e.message === 'Abort!' ? undefined : e });
+        return of(e);
+      }),
+      finalize(() => {
+        if (erroredOrAborted && uploadTarget) {
+          this.client.delete({ id: uploadTarget.id, kind: this.fileKind.id });
+        }
+      })
+    );
+  };
+
+  public upload = (meta?: unknown): Observable<void> => {
+    if (this.isUploading()) {
+      throw new Error('Upload already in progress');
+    }
+    this.uploading$.next(true);
```

Comment:
Maybe more correct would be to derive this `uploading$` state from the actual file state, that way it would also be correct if uploads could be started multiple times:

```ts
constructor () {
  this.uploads$
    .pipe(
      switchMap(uploads => combineLatest(uploads.map(upload => upload.status$))),
      map(statuses => {
        for (const status of statuses) if (status === 'uploading') return true;
        return false.
      }),
    )
    .subscribe(this.uploading$);
 
}
```